Objects/listobject.c

```diff
@@ -522,49 +522,47 @@ list_dealloc(PyObject *self)
 static PyObject *
 list_repr_impl(PyListObject *v)
 {
-    PyObject *s;
-    _PyUnicodeWriter writer;
-    Py_ssize_t i = Py_ReprEnter((PyObject*)v);
-    if (i != 0) {
-        return i > 0 ? PyUnicode_FromString("[...]") : NULL;
+    int res = Py_ReprEnter((PyObject*)v);
+    if (res != 0) {
+        return (res > 0 ? PyUnicode_FromString("[...]") : NULL);
     }
 
-    _PyUnicodeWriter_Init(&writer);
-    writer.overallocate = 1;
     /* "[" + "1" + ", 2" * (len - 1) + "]" */
-    writer.min_length = 1 + 1 + (2 + 1) * (Py_SIZE(v) - 1) + 1;
+    Py_ssize_t prealloc = 1 + 1 + (2 + 1) * (Py_SIZE(v) - 1) + 1;
+    PyUnicodeWriter *writer = PyUnicodeWriter_Create(prealloc);
+    if (writer == NULL) {
+        goto error;
+    }
 
-    if (_PyUnicodeWriter_WriteChar(&writer, '[') < 0)
+    if (PyUnicodeWriter_WriteChar(writer, '[') < 0) {
         goto error;
+    }
 
     /* Do repr() on each element.  Note that this may mutate the list,
        so must refetch the list size on each iteration. */
-    for (i = 0; i < Py_SIZE(v); ++i) {
+    for (Py_ssize_t i = 0; i < Py_SIZE(v); ++i) {
         if (i > 0) {
-            if (_PyUnicodeWriter_WriteASCIIString(&writer, ", ", 2) < 0)
+            if (PyUnicodeWriter_WriteUTF8(writer, ", ", 2) < 0) {
```

Comment:
What is faster, `PyUnicodeWriter_WriteUTF8(writer, ", ", 2)` or two `PyUnicodeWriter_WriteChar()`?

This is an idle question. Even if the latter is marginally faster, it may still be more preferable to use the former. But if the difference is significant, this may be a question for optimization.

Reply:
> What is faster, PyUnicodeWriter_WriteUTF8(writer, ", ", 2) or two PyUnicodeWriter_WriteChar()?

Two PyUnicodeWriter_WriteChar() calls is faster:

```
+----------------+-------------+-----------------------+
| Benchmark      | change_utf8 | change_2char          |
+================+=============+=======================+
| repr([])       | 50.2 ns     | 54.0 ns: 1.07x slower |
+----------------+-------------+-----------------------+
| repr(list3)    | 825 ns      | 802 ns: 1.03x faster  |
+----------------+-------------+-----------------------+
| repr(list4)    | 2.89 us     | 2.66 us: 1.09x faster |
+----------------+-------------+-----------------------+
| Geometric mean | (ref)       | 1.01x faster          |
+----------------+-------------+-----------------------+
```